### ISSUE TITLE
[Python] Segfault when filtering table with fixed_size_binary

### ISSUE THREAD

**Issue by rabernat** [apache/arrow#43236]
### Describe the bug, including details regarding any error messages, version, and platform.

I want to filter a pyarrow table based on a fixed-length tuple of ints. Filtering is not supported for compound dtypes like list arrays, so I am converting this tuple to a custom binary encoding and filtering against that.

I am experiencing a segfault which occurs only when:
- the data are sufficiently large (in this example we have 1_000_000 rows)
- I use a `fixed_size_binary` type rather than `binary`

Here is a MRE

```python
from itertools import product
import pyarrow.compute as pc
import pyarrow as pa

def encode_coord_as_fixed_size_binary(coord: tuple[int, ...], max_length=2) -> bytes:
    """Turn a tuple of ints into bytes"""
    out = b''
    for item in coord[::-1]:
        out += item.to_bytes(length=max_length)
    assert len(out) == len(coord) * max_length
    return out

N = 100  # works fine for smaller data (e.g. N=10)
nrows = N**3

raw_data = list(product(range(N), repeat=3))
encoded_data = [encode_coord_as_fixed_size_binary(coord) for coord in raw_data]
array = pa.array(encoded_data, type=pa.binary())
array_fixed_size = pa.array(encoded_data, pa.binary(3 * 2))

table = pa.table([array, array_fixed_size], names=["binary", "fixed_size_binary"])

# works fine
table.filter(pc.field("binary") == encode_coord_as_fixed_size_binary((5, 3, 9)))

# this segfaults
table.filter(pc.field("fixed_size_binary") == encode_coord_as_fixed_size_binary((5, 3, 9)))
```

There is a core dumped, but it is like 500 MB, so I'm not including it here.

I'm on:
- Linux
- Python 3.12.4
- PyArrow 16.1.0

### Component(s)

Python

**Comment by zanmato1984** [comment]
Hi, thanks reporting.

I was able to repro the issue using conda pyarrow 16.1.0 (however I failed to do so with my own built arrow/pyarrow). Attaching the stack trace here:
```
(lldb) bt
* thread #5, stop reason = EXC_BAD_ACCESS (code=1, address=0x1000)
  * frame #0: 0x0000000197967230 libsystem_platform.dylib`_platform_memmove + 144
    frame #1: 0x0000000103d7f10c libarrow.1601.0.0.dylib`void arrow::internal::TransferBitmap<(arrow::internal::TransferMode)0>(unsigned char const*, long long, long long, long long, unsigned char*) + 216
    frame #2: 0x0000000103d7fbc0 libarrow.1601.0.0.dylib`arrow::Result<std::__1::shared_ptr<arrow::Buffer>> arrow::internal::TransferBitmap<(arrow::internal::TransferMode)0>(arrow::MemoryPool*, unsigned char const*, long long, long long) + 144
    frame #3: 0x00000001031eb710 libarrow.1601.0.0.dylib`std::__1::enable_if<std::is_same<arrow::FixedSizeBinaryType, arrow::FixedSizeBinaryType>::value && !std::is_same<arrow::BinaryType, arrow::FixedSizeBinaryType>::value, arrow::Status>::type arrow::compute::internal::(anonymous namespace)::BinaryToBinaryCastExec<arrow::BinaryType, arrow::FixedSizeBinaryType>(arrow::compute::KernelContext*, arrow::compute::ExecSpan const&, arrow::compute::ExecResult*) + 368
    frame #4: 0x0000000103141aec libarrow.1601.0.0.dylib`arrow::compute::detail::(anonymous namespace)::ScalarExecutor::Execute(arrow::compute::ExecBatch const&, arrow::compute::detail::ExecListener*) + 1628
    frame #5: 0x000000010314b7d4 libarrow.1601.0.0.dylib`arrow::compute::ExecuteScalarExpression(arrow::compute::Expression const&, arrow::compute::ExecBatch const&, arrow::compute::ExecContext*) + 1652
    frame #6: 0x000000010314b550 libarrow.1601.0.0.dylib`arrow::compute::ExecuteScalarExpression(arrow::compute::Expression const&, arrow::compute::ExecBatch const&, arrow::compute::ExecContext*) + 1008
    frame #7: 0x000000013156e9b0 libarrow_acero.1601.0.0.dylib`arrow::acero::(anonymous namespace)::FilterNode::ProcessBatch(arrow::compute::ExecBatch) + 212
    frame #8: 0x0000000131597c50 libarrow_acero.1601.0.0.dylib`arrow::acero::MapNode::InputReceived(arrow::acero::ExecNode*, arrow::compute::ExecBatch) + 204
    frame #9: 0x00000001315c1fd8 libarrow_acero.1601.0.0.dylib`std::__1::__function::__func<arrow::acero::(anonymous namespace)::SourceNode::SliceAndDeliverMorsel(arrow::compute::ExecBatch const&)::'lambda'(), std::__1::allocator<arrow::acero::(anonymous namespace)::SourceNode::SliceAndDeliverMorsel(arrow::compute::ExecBatch const&)::'lambda'()>, arrow::Status ()>::operator()() + 364
    frame #10: 0x00000001315a4fc4 libarrow_acero.1601.0.0.dylib`std::__1::enable_if<!std::is_void<arrow::Status>::value && !is_future<arrow::Status>::value && (!arrow::Future<arrow::internal::Empty>::is_empty || std::is_same<arrow::Status, arrow::Status>::value), void>::type arrow::detail::ContinueFuture::operator()<std::__1::function<arrow::Status ()>&, arrow::Status, arrow::Future<arrow::internal::Empty>>(arrow::Future<arrow::internal::Empty>, std::__1::function<arrow::Status ()>&) const + 60
    frame #11: 0x00000001315a4ef0 libarrow_acero.1601.0.0.dylib`std::__1::__bind_return<arrow::detail::ContinueFuture, std::__1::tuple<arrow::Future<arrow::internal::Empty>, std::__1::function<arrow::Status ()>>, std::__1::tuple<>, __is_valid_bind_return<arrow::detail::ContinueFuture, std::__1::tuple<arrow::Future<arrow::internal::Empty>, std::__1::function<arrow::Status ()>>, std::__1::tuple<>>::value>::type std::__1::__bind<arrow::detail::ContinueFuture, arrow::Future<arrow::internal::Empty>&, std::__1::function<arrow::Status ()>>::operator()[abi:fe180100]<>() + 68
    frame #12: 0x0000000103dde024 libarrow.1601.0.0.dylib`void* std::__1::__thread_proxy[abi:fe180100]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, arrow::internal::ThreadPool::LaunchWorkersUnlocked(int)::$_0>>(void*) + 496
    frame #13: 0x0000000197936f94 libsystem_pthread.dylib`_pthread_start + 136
```

**Comment by zanmato1984** [comment]
I have verified that this issue is fixed in upcoming 17.0 release by PR #41452.

@rabernat Would you like to give it a try? If you have any trouble getting/building the latest binary, you can also wait for the official release of arrow 17.0 which will come in a couple of days.

Thanks.

**Comment by zanmato1984** [comment]
I'm closing this as a dup of #41306. Please feel free to reopen if you ran into this issue again on arrow 17.0. Thanks.
